### PR TITLE
Reset repository state before pull

### PR DESCRIPTION
**Pull request by JuneKelly**
Part of the fix for https://github.com/overleaf/issues/issues/3415

While investigating the issue, we found that the git-bridge leaves its repos in a "dirty" state after each operation, by removing all repo files from disk. This is usually not a problem.

However, in the case where a push has been made, the repo is left in a state like this:

```
root@1ca5aac0c601:/tmp/wlgb/5f6b092958a3bd00260d19f3# git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)
	deleted:    test.tex
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
	deleted:    main.tex
```

Where all files are deleted, but the newly pushed file is _staged_ for deletion. I don't know why this happens exactly.

When a pull is made, we fetch the project content from web, and splat it into the repository, and run (effectively) `git add .` , then commit the changes. This `add` operation takes a custom tree-walker object which ignores the usual git-ignore imperatives, and so always adds all files coming from web. Then we commit the changes and stream the content back to the git client.

However, this git-ignore-ignore strategy only seems to work on `.gitignore` specs at the top level of the project. This brings us to the original issue, that when a `.gitignore` exists in a sub-directory, and a file has been force-added, this file can be _pushed_ to the git-bridge, but will then be erased on the next _pull_.

This happens because the file is already staged for deletion, before the pull operation even starts. and because the file is not added (due to gitignore), and is staged for deletion, the operation completes as a delete.

(well, to be fair, that's one problem amongst many, but it seems to be the problem closest to the root of the Tree of Problems)

This PR changes the _pull_ path so that we do a `git reset --hard HEAD` before writing the content from web, which ensures that these lingering staged deletes don't hang around and mess with subsequent operations.

Arguably, we should also either make the git-ignore-ignore-tree-walker recursive, or find a way to pass the `-f` flag to the `add` operation (which I could not see a way to do in the JGit API), but this change is enough to solve the current problem.

(Note: we seem to have a kind of asymmetry between the push and pull paths. On the _push_ side, JGit writes the incoming changes to the `.git` folder in the repo before any of our code even runs, while on the _pull_ side we run our code and then write the changes to the repo with `git add; git commit` )

### CODE REVIEW

**Comment by gh2k** [comment]
I've unlinked this PR from https://github.com/overleaf/issues/issues/3415 (by removing the `fixes` keyword) because I don't think it will fix 100% of the bug. Specifically, when a file is added at the Overleaf end which matches a `.gitignore` file in a subdirectory, I don't think it will be added to the repo. - I think that this works as a PR on its own though, and that can be done in a follow-up.

**Review by gh2k — APPROVED** [review]
creepysmile

**Comment by JuneKelly** [comment]
@gh2k I've added a few commits to solve the pull-ignored-file-from-overleaf problem :+1:

File: src/main/java/uk/ac/ic/wlgitbridge/bridge/repo/NoGitignoreIterator.java [hunk 1]
```diff
@@ -74,4 +77,11 @@ protected void init(Entry[] list) {
         }
     }
 
+    // When entering a sub-directory, create a new instance of this class,
```
Comment (gh2k) [inline src/main/java/uk/ac/ic/wlgitbridge/bridge/repo/NoGitignoreIterator.java:80]: :tada: nice

**Review by gh2k — APPROVED** [review]
Nice one!
approval

Thanks for adding that. :+1:

**Pull request merged**